core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -474,16 +480,15 @@ public synchronized void initialize(URI uri, org.apache.hadoop.conf.Configuratio
 
     // HDFS doesn't allow the authority to be empty; it must be "/" instead.
     String authority = uri.getAuthority() == null ? "/" : uri.getAuthority();
-    mAlluxioHeader = getScheme() + "://" + authority;
+    // Use URI's scheme for shim mode.
+    String scheme = getScheme();
+    // Extract the scheme from URI if FS implementation is not bound to a scheme.
+    if (scheme.equals(Constants.NO_SCHEME)) {
```

Comment:
It looks to me we are testing which class we are (`FileSystem` or `ShimFileSystem`), and then make decision to create `mAlluxioHeader`.  If this is the case, I think it will cleaner to move this logic further to the implementing class rather in the base Abstract class. In other words, the logic of the base Abstract class should not depend on `Constants.NO_SCHEME` which is the local information in the implementing class `ShimFileSystem`

Reply:
Agreed.